### DISCUSSION TITLE
[API Proposal]: Auto-infer `Span<T>` type for stackalloc in safe contexts.

### DISCUSSION THREAD

**Discussion by DrkWzrd** [dotnet/csharplang#6025] [Language Ideas] [answered]
### Background and motivation

In advance, sorry for my english. An sorry if this is not the right repo.

The title is self-explanatory.

When we are in safe context we have to write `Span<T> span = stackalloc T[number];` to avoid the error. That's because (obviously) stackalloc creates a `T*` element and, "hey friend, are you sure what are you doing?".

But it would be so hard to make the compiler assume when we are in a safe context

### API Proposal

```C#
public void ThisIsSafe(){
    var span = stackalloc double[20]; //"Oh this is safe, then I'll treat this like a Span for the merely beginning.
}

public unsafe void ThisIsUnsafe(){
   var pointer = stackalloc double[20]; //A pointer, everything is OK, continue.
}
```

### API Usage

API proposal.

### Alternative Designs

No one.

### Risks

I can't see anyone, but sure this hasn't been done for a good reason.

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-memory
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

### Background and motivation

In advance, sorry for my english. An sorry if this is not the right repo.

The title is self-explanatory.

When we are in safe context we have to write `Span<T> span = stackalloc T[number];` to avoid the error. That's because (obviously) stackalloc creates a `T*` element and, "hey friend, are you sure what are you doing?".

But it would be so hard to make the compiler assume when we are in a safe context

### API Proposal

```C#
public static void ThisIsSafe(){
    var span = stackalloc double[20]; //"Oh this is safe, then I'll treat this like a Span for the merely beginning.
}

public static unsafe void ThisIsUnsafe(){
   var pointer = stackallock double[20]; //A pointer, everything is OK, continue.
}
```


### API Usage

API proposal.

### Alternative Designs

No one.

### Risks

I can't see anyone, but sure this hasn't been done for a good reason.

<table>
  <tr>
    <th align="left">Author:</th>
    <td>DrkWzrd</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`api-suggestion`, `area-System.Memory`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by mburbea** [comment, marked as answer]
As a workaround you can do:
```cs
public void SafeMethod(){
     var span = (stackalloc double[20]);
}
```
Wrapping it in parenthesis changes the inferred type to `Span<T>` .

**Reply by DrkWzrd** [reply]
I didn't know that.

Thanks a lot. I'll use it that way for now on although I still think parethensis shouldn't be needed, but I understand that for 2 keystrokes the needed changes doesn't worth the time.

Compiler and analyzers can infer the safety-unsafety of the code, isn't it?

Anyway, I mark this as the answer.

**Reply by DrkWzrd** [reply]
Another question I think right now. Does this subtle change change in any form the generated JIT code? Just curiosity.
